WIP Build two arch version 2

tpatki opened the pull request:
# Description

Rebasing Stephanie's existing PR #121 was hard due to several conflicts and changes in variorum, so this PR is a fresh start at adding her changes on v0.6.0, and then further add support for AMD CPU, GPU, Intel GPU, JSON, docs, etc.. 

Fixes #62.
Supersedes #121 .

ToDo:

- [x] Add AMD CPU support
- [x] Fix AMD build error with DENABLE_WARNINGS=On (submit separate PR)
- [x] Update JSON APIs to be node-level and not component level (similar to best effort node power capping), node-level APIs are not supported on GPU-only builds. The GPU values should be populated internally through a different GPU-specific API at some point in a different PR. 

~~- [ ] Update JSON APIs to support multiple-platforms~~ <- separate PR as we need to add JSON API for GPUs so it can populate the JSON string at node level. 

- [x] Update the best_effort-node-power-limit code to maintain its v0.6.0 release behavior, it is a node-level API not a component-level (CPU, Mem, GPU level) API. 
- [x] Set a MAX_PLATFORMS variable so we can easily update in the future if we have a situation where we are building for a CPU + GPU + Other Accelerator situation.

~- [ ] Update documentation to reflect how to build for CPU + GPU, and how we only support 2 platforms at a time~ <-- separate docs PR 

- [x] (Start a separate discussion about how to handle component power capping and what "best effort" means when GPU power capping is available)

## Type of change

- [x] New feature/architecture support (non-breaking change which adds functionality)
- [ ] Documentation update
- [x] Build/CI update

# How Has This Been Tested?

Tested build with -DENABLE_WARNINGS=On,  as well as variorum-print-power-example, variorum-cap-best-effort-node-power-cap-example and variorum-get-node-power-json-example wherever applicable.

_Note that the best-effort-power-cap and get-node-power-json examples are node-level, and will only work on the primary platform component which we assume to currently be the CPU. Future goal includes adding support for obtaining detailed GPU information for the node-level APIs -- current behavior of variorum is maintained in this PR for these node-level functions.._ 

- [x] Comus single platform
- [x] Rhetoric single platform
- [x] Thompson single platform
- [x] Juno single platform
- [x] Corona GPU single platform
- [x] Lassen CPU single platform
- [x] Lassen GPU single platform
- [x] Lassen CPU + GPU (two platforms): print power, node-level json and best effort cap works as expected. Best effort example output is confusing as when printing, NVML power limit doesn't show as changed even though it has (a derived power cap is indirectly chosen by IBM to set through IBM's node limit, which supersedes any NVML limits.).
- [x] Tioga CPU single platform
- [x] Tioga GPU single platform
- [x] Tioga CPU+GPU single platform: Similar to Lassen, print power and print power limit and JSON APIs work as expected. Best effort power cap cannot be tested here due to lack of permissions. 

# Checklist:

- [x] I have run `./scripts/check-code-format.sh` and confirm my C/C++ code follows the style guidelines of variorum
- [ ] I have run `flake8` and `black --check --diff` and confirm my python code follows the style guidelines of variorum
- [ ] I have run `./scripts/check-rst-format.sh` and confirm my documentation follows the style guidelines of variorum
- [x] I have added comments in my code
- [x] My changes generate no new warnings (build with `-DENABLE_WARNINGS=ON`)
- [x] New and existing unit tests pass with my changes

Thank you for taking the time to contribute to Variorum!

tpatki committed:
Add hostconfig file for building for two arch.

tpatki committed:
Updates to config_arch.h to support building for multiple arch.

tpatki committed:
WIP: need to understand existing PR more to ensure changes in variorum enter/exit are correct. To be picked up on Monday.

tpatki committed:
Intel file changes as picked up from Stephanie's branch

tpatki committed:
ARM changes as picked up from Stephanie's branch

tpatki committed:
MSR changes as picked up from Stephanie's branch

tpatki committed:
IBM changes as picked up from Stephanie's branch

tpatki committed:
Add P_NVIDIA_GPU_IDX, replace P_INTEL_IDX with P_INTEL_CPU_IDX.

tpatki committed:
Relace P_IBM with P_IBM_CPU

tpatki committed:
Add missing P_ARM_CPU_IDX

tpatki committed:
Add support for Intel DGPU

tpatki committed:
Update config_arch.c

tpatki committed:
Format files

tpatki committed:
Fix several build errors with respect to unsigned vs signed int, and missing P_INTEL_CPU_IDX

tpatki committed:
Pick variorum changes from Stephanie's branch

tpatki committed:
Fix more signed vs unsigned errors in msr_core.c, comment out the best effort node power limit for now, because we want that to be node level.

tpatki committed:
Formatting fixes

tpatki committed:
Ignore Best Effort Node Capping for now

tpatki committed:
translate function should have P_INTEL_CPU_IDX as the idx parameter input, not a hard-coded value of 0 for idx.

tpatki committed:
Formatting fixes

tpatki committed:
Add a P_MSR_CORE_IDX to track common MSR functionality across Intel and AMD codebases. Set MSR_CORE_IDX based on Intel or AMD CPU correctly in detect_arch.

tpatki committed:
Add a P_MSR_CORE_IDX to track common MSR functionality across Intel and AMD codebases. Set MSR_CORE_IDX based on Intel or AMD CPU correctly in detect_arch.

tpatki committed:
Fix build error around MSR_CORE_IDX by using extern declaration

tpatki committed:
Add AMD GPU support

tpatki committed:
Add max platforms, restore behavior of best effort power limit to what it was before the multi-arch build

tpatki committed:
Add ifdef instead of if statements as platform IDs are defined based on ifdefs.

tpatki committed:
Remove older continue statement

tpatki committed:
Fix stupid copy-paste error

tpatki committed:
Update the host config for CPU+GPU build for lassen

tpatki committed:
Initial support for AMD for multi platform. Untested

tpatki committed:
Fix AMD multi-platform build errors

tpatki committed:
Add missing parameter for AMD GPU multi-platform build

tpatki committed:
Add missing ARM idx check

tpatki committed:
Fix missed re-definition from rebase

tpatki committed:
JSON APIs are node-level and not per-component, make relevant changes

tpatki committed:
Add missing index for ARM JSON API

slabasan commented on src/variorum/variorum.h:
```diff
@@ -226,7 +226,6 @@ int variorum_print_verbose_thermals(void);
 /// - Intel Kaby Lake
 /// - Intel Cascade Lake
 /// - Intel Cooper Lake
-/// - Intel Arctic Sound
```
Why are these being removed?

tpatki replied:
That looks like a mistake, let me revert that shortly.

tpatki replied:
Actually, not a mistake, @slabasan . Intel Arctic Sound is a GPU, and doesn't support these APIs (verbose ones) in the port.

slabasan commented on src/variorum/config_architecture.h:
```diff
@@ -268,7 +282,8 @@ int variorum_exit(const char *filename,
                            unsigned *ncores,
-                           unsigned *nthreads);
+                           unsigned *nthreads,
+                           int idx);
```
what is this index being used for?

tpatki replied:
The number of sockets, threads etc and the hostname is set in the variorum_get_topology function in config_architecture.c, which still needs to be done per platform.

slabasan replied:
I'm seeing that sometimes `P_MSR_CORE_IDX` is being passed into `variorum_get_topology`, what does this index represent?

tpatki replied:
`MSR_CORE_IDX` was added for scenarios where we need to share the `msr_core.c` functionality -- especially among Intel and AMD CPUs. Previously, `msr_core.c` was using the `P_INTEL_CPU_IDX` internally, which will not work correctly if we have an AMD CPU platform, which would need the `P_AMD_CPU_IDX` instead. So, I added the `MSR_CORE_IDX`, which can now be set based on the platform that we're building for, and be passed to common internal interfaces such as the ones from `msr_core.c`.

slabasan committed:
expand tab to spaces

slabasan reviewed (dismissed):
Build looks good to me, will you be adding a host config file for the tioga dual build platform as well?

tpatki commented:
@slabasan Yes will add tioga host configs shortly. Was thinking of a separate PR, need to create a common installation like we discussed -- would you like me to just add it to this PR?

tpatki committed:
Clean up print statements, revert Arctic Sound related change.

tpatki commented:
@slabasan : as discussed, reverted the change around Arctic Sound (will submit separate PR), and removed the `<Inte/AMD> is idx` prints, and moved the number of registered platforms print to the variorum log instead. Let me know if I missed anything, I believe this is ready to go. Thank you!
